### PR TITLE
[BTAT-5226] - added PendingDeregPredicate and tests

### PR DESCRIPTION
**Pull request by sraspin**

### CODE REVIEW

File: app/controllers/DeregistrationConfirmationController.scala [hunk 1]
```diff
@@ -18,20 +18,21 @@ package controllers
...
 class DeregistrationConfirmationController @Inject()(val messagesApi: MessagesApi,
                                                      val authentication: AuthPredicate,
+                                                     val pendingDeregCheck: PendingChangesPredicate,
                                                      val deleteAllStoredAnswersService: DeleteAllStoredAnswersService,
                                                      val serviceErrorHandler: ServiceErrorHandler,
                                                      val customerDetailsService: CustomerDetailsService,
                                                      implicit val appConfig: AppConfig) extends FrontendController with I18nSupport {
 
-  val show: Action[AnyContent] = authentication.async { implicit user =>
+  val show: Action[AnyContent] = (authentication andThen pendingDeregCheck).async { implicit user =>
```
Comment (TomDudley92) [inline app/controllers/DeregistrationConfirmationController.scala]: Would this need the new predicate one? It displays once the pending change has gone through so in theory you wouldn't be able to access this page with this predicate here

File: app/controllers/predicates/PendingChangesPredicate.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.predicates
+
+import config.{AppConfig, ServiceErrorHandler}
+import common.SessionKeys.pendingDeregKey
+import javax.inject.Inject
+import models.{ChangeIndicatorModel, ErrorModel, PendingDeregModel, User}
```
Comment (TomDudley92) [inline app/controllers/predicates/PendingChangesPredicate.scala]: Can you remove the unused imports please

File: app/controllers/predicates/PendingChangesPredicate.scala [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  private def getCustomerInfoCall[A](vrn: String)(implicit hc: HeaderCarrier,
+                                                  request: User[A]): Future[Either[Result, User[A]]] =
+    customerDetailsService.getDeregPending(vrn).map {
+      case Right(pending) =>
+        pending.deregistration match {
+          case Some(PendingDeregModel(true)) =>
+            Logger.warn("[PendingChangesPredicate][getCustomerInfoCall] - " +
+              "Deregistration pending. Rendering graceful error page.")
+            Left(serviceErrorHandler.showInternalServerError.addingToSession(pendingDeregKey -> "true"))
```
Comment (TomDudley92) [inline app/controllers/predicates/PendingChangesPredicate.scala]: This isn't a graceful error page. Is there supposed to be one here or this is being dealt with in a later story?

File: test/controllers/CannotDeregisterThresholdControllerSpec.scala [hunk 4]
```diff
@@ -25,7 +25,12 @@ import scala.concurrent.Future
 
 class CannotDeregisterThresholdControllerSpec extends ControllerBaseSpec {
 
-  object TestCannotDeregisterThresholdController extends CannotDeregisterThresholdController(messagesApi, mockAuthPredicate, mockConfig)
+  object TestCannotDeregisterThresholdController extends CannotDeregisterThresholdController(
```
Comment (TomDudley92) [inline test/controllers/CannotDeregisterThresholdControllerSpec.scala]: Can you remove the unused imports in this file please

File: test/controllers/TaxableTurnoverControllerSpec.scala [hunk 5]
```diff
@@ -34,6 +34,7 @@ class TaxableTurnoverControllerSpec extends ControllerBaseSpec with MockWipeRedu
   object TestTaxableTurnoverController extends TaxableTurnoverController(
     messagesApi,
     mockAuthPredicate,
+    mockPendingDeregPredicate,
```
Comment (TomDudley92) [inline test/controllers/TaxableTurnoverControllerSpec.scala]: Can you remove the unused imports in the file please

File: test/mocks/MockAuth.scala [hunk 6]
```diff
@@ -112,5 +114,20 @@ trait MockAuth extends TestUtil with MockFactory {
     )
   )
 
+  val mockPendingDeregPredicate: PendingChangesPredicate = {
```
Comment (TomDudley92) [inline test/mocks/MockAuth.scala]: This isn't part of Auth. Can it moved to it's own file please

File: it/stubs/AuthStub.scala [hunk 7]
```diff
@@ -63,6 +63,31 @@ object AuthStub extends WireMockMethods {
       .thenReturn(status = UNAUTHORIZED, headers = Map("WWW-Authenticate" -> """MDTP detail="MissingBearerToken""""))
   }
 
+  def noDeregPending(): StubMapping = {
```
Comment (TomDudley92) [inline it/stubs/AuthStub.scala]: This shouldn't be part of the auth stub really. Can we have a VatSubscription created please

File: it/pages/WhyTurnoverBelowISpec.scala [hunk 8]
```diff
@@ -30,7 +30,7 @@ class WhyTurnoverBelowISpec extends IntegrationBaseSpec {
 
   "Calling the GET Why Turnover Below endpoint" when {
 
-    def getRequest(): WSResponse = get("/reasons-for-low-turnover")
+    def getRequest(): WSResponse = get("/reasons-for-low-turnover", formatPendingDereg(Some("false")))
```
Comment (TomDudley92) [inline it/pages/WhyTurnoverBelowISpec.scala]: You don't need parenthesis on this method

File: it/pages/VATAccountsISpec.scala [hunk 9]
```diff
@@ -31,7 +31,7 @@ class VATAccountsISpec extends IntegrationBaseSpec {
 
   "Calling the GET VatAccounts" when {
 
-    def getRequest(): WSResponse = get("/accounting-method")
+    def getRequest(): WSResponse = get("/accounting-method", formatPendingDereg(Some("false")))
```
Comment (TomDudley92) [inline it/pages/VATAccountsISpec.scala]: You don't need parenthesis on this method

File: it/pages/OutstandingInvoicesISpec.scala [hunk 10]
```diff
@@ -31,7 +31,7 @@ class OutstandingInvoicesISpec extends IntegrationBaseSpec {
 
     val validYesModel = Yes
 
-    def getRequest(): WSResponse = get("/outstanding-invoices")
+    def getRequest(): WSResponse = get("/outstanding-invoices", formatPendingDereg(Some("false")))
```
Comment (TomDudley92) [inline it/pages/OutstandingInvoicesISpec.scala]: You don't need parenthesis on this method

File: it/pages/IssueNewInvoicesISpec.scala [hunk 11]
```diff
@@ -31,7 +31,7 @@ class IssueNewInvoicesISpec extends IntegrationBaseSpec {
 
   "Calling the GET IssueNewInvoices" when {
 
-    def getRequest(): WSResponse = get("/new-invoices")
+    def getRequest(): WSResponse = get("/new-invoices", formatPendingDereg(Some("false")))
```
Comment (TomDudley92) [inline it/pages/IssueNewInvoicesISpec.scala]: You don't need parenthesis on this method

File: it/pages/DeregistrationReasonISpec.scala [hunk 12]
```diff
@@ -29,9 +29,9 @@ import stubs.DeregisterVatStub
 
 class DeregistrationReasonISpec extends IntegrationBaseSpec {
 
-  "Calling the GET Why Turnover Below endpoint" when {
+  "Calling the GET Deregistration reason endpoint" when {
 
-    def getRequest(): WSResponse = get("/deregister-reason?isAgent=false")
+    def getRequest(): WSResponse = get("/deregister-reason?isAgent=false", formatPendingDereg(Some("false")))
```
Comment (TomDudley92) [inline it/pages/DeregistrationReasonISpec.scala]: You don't need parenthesis on this method

File: it/pages/DeregistrationDateISpec.scala [hunk 13]
```diff
@@ -39,7 +39,7 @@ class DeregistrationDateISpec extends IntegrationBaseSpec {
 
   "Calling the GET Deregistration Date endpoint" when {
 
-    def getRequest(): WSResponse = get("/deregister-date")
+    def getRequest(): WSResponse = get("/deregister-date", formatPendingDereg(Some("false")))
```
Comment (TomDudley92) [inline it/pages/DeregistrationDateISpec.scala]: You don't need parenthesis on this method

File: it/pages/CheckYourAnswersISpec.scala [hunk 14]
```diff
@@ -23,13 +23,13 @@ import services._
...
 import play.api.libs.ws.WSResponse
-import stubs.DeregisterVatStub
+import stubs.{AuthStub, DeregisterVatStub}
 
 class CheckYourAnswersISpec extends IntegrationBaseSpec {
 
   "Calling GET Check your answers" when {
 
-    def getRequest(): WSResponse = get("/check-your-answers")
+    def getRequest(): WSResponse = get("/check-your-answers", formatPendingDereg(Some("false")))
```
Comment (TomDudley92) [inline it/pages/CheckYourAnswersISpec.scala]: You don't need parenthesis on this method

File: it/pages/CheckYourAnswersISpec.scala [hunk 15]
```diff
@@ -23,13 +23,13 @@ import services._
 import play.api.http.Status._
 import play.api.libs.json.Json
 import play.api.libs.ws.WSResponse
-import stubs.DeregisterVatStub
+import stubs.{AuthStub, DeregisterVatStub}
```
Comment (TomDudley92) [inline it/pages/CheckYourAnswersISpec.scala]: Can you remove the unused import please

File: app/controllers/DeregistrationConfirmationController.scala (hunk 1 above)
Comment (sraspin) [reply app/controllers/DeregistrationConfirmationController.scala]: Hi pal, I think it's better to keep it here. If the user has gone through the journey as they should then the sessionKey would already be set so the predicate wouldn't make another call to vat-subscription.

Comment (TomDudley92) [reply app/controllers/DeregistrationConfirmationController.scala]: But the user would have a pending change as they've just made one and wouldn't be able to view the confirmation page therefore we don't need the predicate on this page

File: test/connectors/httpParsers/PendingDeregHttpParserSpec.scala [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    "the http response status is OK with valid Json" should {
+
+      "return a CustomerDetailsModel" in {
+        PendingDeregReads.read("", "", HttpResponse(Status.OK, Some(pendingDeregFalseJson))) shouldBe Right(pendingDeregFalse)
+      }
+    }
+
+    "the http response status is OK with invalid Json" should {
+
+      "return a 'ChangeIndicatorModel(None)'" in {
```
Comment (TomDudley92) [inline test/connectors/httpParsers/PendingDeregHttpParserSpec.scala]: So this is a correct scenario but there still needs to be a scenario for invalid json being retrieved

**Pull request merged**